### PR TITLE
Add `getOfflineRegion` Kotlin API

### PR DESCRIPTION
**Pull request by JRWilding**
Adds getOfflineRegion to the kotlin api for android so that we don't have to get the list of all regions and iterate through them.

### CODE REVIEW

File: platform/default/src/mbgl/storage/offline_database.cpp [hunk 1]
```diff
@@ -835,6 +835,25 @@ expected<OfflineRegions, std::exception_ptr> OfflineDatabase::listRegions() try
...
 
+expected<OfflineRegion, std::exception_ptr> OfflineDatabase::getRegion(const int64_t regionID) try {
+    mapbox::sqlite::Query query{getStatement("SELECT definition, description FROM regions WHERE id = ?1")};
+    query.bind(1, regionID);
+    query.run();
+
+    const auto definition = query.get<std::string>(0);
+    const auto description = query.get<std::vector<uint8_t>>(1);
+
+    if (definition.empty() and description.empty()) {
```
Comment (louwers) [inline platform/default/src/mbgl/storage/offline_database.cpp:846]: Somehow Qt is not happy with this `and`...

Maybe just use `&&`.

File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/offline/OfflineManager.kt [hunk 2]
```diff
@@ -60,6 +60,27 @@ class OfflineManager private constructor(context: Context) {
...
+     * OfflineRegion in the database or an error message otherwise.
+     */
+    @Keep
+    interface GetOfflineRegionCallback {
+        /**
+         * Receives the offline region.
+         *
+         * @param offlineRegion the offline region
+         */
+        fun onRegion(offlineRegion: OfflineRegion)
```
[the remark this answers is not in the document]
Comment (JRWilding) [reply platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/offline/OfflineManager.kt:74]: How about `Result<OfflineRegion?>`  just null if not found?

[the remark this answers is not in the document]
Comment (louwers) [reply platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/offline/OfflineManager.kt:74]: @JRWilding Sure, that works! I just did a quick Google search and Stack Overflow seems to agree that is better.

[the remark this answers is not in the document]
Comment (JRWilding) [reply platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/offline/OfflineManager.kt:74]: I went with adding a onRegionNotFound to not introduce Result to make it more consistent with the other apis

File: platform/android/MapLibreAndroid/src/cpp/offline/offline_manager.cpp [hunk 3]
```diff
@@ -63,6 +63,31 @@ void OfflineManager::listOfflineRegions(jni::JNIEnv& env_,
...
+    fileSource->getOfflineRegion(
+        regionID,
+        [
+            // Keep a shared ptr to a global reference of the callback and file
+            // source so they are not GC'd in the meanwhile
+            callback = std::make_shared<decltype(globalCallback)>(std::move(globalCallback)),
+            jFileSource = std::make_shared<decltype(globalFilesource)>(std::move(globalFilesource))](
+            mbgl::expected<mbgl::OfflineRegion, std::exception_ptr> region) mutable {
+            // Reattach, the callback comes from a different thread
+            android::UniqueEnv env = android::AttachEnv();
```
[the remark this answers is not in the document]
Comment (JRWilding) [reply platform/android/MapLibreAndroid/src/cpp/offline/offline_manager.cpp:82]: I'm not exactly sure how to applies this and the above suggestion. I've kept these consistent with the other methods in the file.

[the remark this answers is not in the document]
Comment (louwers) [reply platform/android/MapLibreAndroid/src/cpp/offline/offline_manager.cpp:82]: Sorry, I misunderstood.

File: platform/android/MapLibreAndroid/src/cpp/offline/offline_manager.cpp [hunk 4]
```diff
@@ -63,6 +63,31 @@ void OfflineManager::listOfflineRegions(jni::JNIEnv& env_,
...
+                                      const jni::Object<GetOfflineRegionCallback>& callback_) {
+    auto globalCallback = jni::NewGlobal<jni::EnvAttachingDeleter>(env_, callback_);
+    auto globalFilesource = jni::NewGlobal<jni::EnvAttachingDeleter>(env_, jFileSource_);
+
+    fileSource->getOfflineRegion(
+        regionID,
+        [
+            // Keep a shared ptr to a global reference of the callback and file
+            // source so they are not GC'd in the meanwhile
+            callback = std::make_shared<decltype(globalCallback)>(std::move(globalCallback)),
```
[the remark this answers is not in the document]
Comment (louwers) [reply platform/android/MapLibreAndroid/src/cpp/offline/offline_manager.cpp:78]: @JRWilding I think you can just do

```cpp
            callback = std::move(globalCallback),
            jFileSource = std::move(globalFilesource)](
```

[the remark this answers is not in the document]
Comment (JRWilding) [reply platform/android/MapLibreAndroid/src/cpp/offline/offline_manager.cpp:78]: I tried this (and various others) with much failure, this is just following the convention in other functions like `listOfflineRegions` and `createOfflineRegion`

File: platform/android/MapLibreAndroid/src/cpp/offline/offline_manager.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (louwers) [reply platform/android/MapLibreAndroid/src/cpp/offline/offline_manager.cpp:82]: Thanks for trying.

**Review by louwers — APPROVED** [review]
I follow [your argument](https://github.com/maplibre/maplibre-native/discussions/2524) to keep the API consistent and Kotlinify it in one full (breaking) swoop. Will merge and make a release with this change if no-one has any qualms about this.

**Comment by louwers** [comment]
After this is merged an issue should be created to add a similar API to iOS.

**Pull request merged**